LibCarla/source/carla/trafficmanager/CollisionStage.cpp

```diff
@@ -166,90 +180,149 @@ namespace CollisionStageConstants {
+    bool ego_stopped_by_light = reference_vehicle_ptr->GetTrafficLightState() != carla::rpc::TrafficLightState::Green;
+    bool ego_at_junction_entrance = !closest_point->CheckJunction() && junction_look_ahead->CheckJunction();
+
+    const ActorId reference_vehicle_id = reference_vehicle->GetId();
+    // Conditions to avoid collision negotiation.
+    if (!(ego_at_junction_entrance && ego_at_traffic_light && ego_stopped_by_light)
+        && !(!ego_inside_junction && !other_vehicle_in_front)
+        && !(!ego_inside_junction && other_vehicle_in_front && !other_vehicle_in_ego_range)
+        && !(ego_inside_junction && !other_vehicles_in_cross_detection_range)
+    ) {
 
       GeometryComparisonCache cache = GetGeometryBetweenActors(reference_vehicle, other_vehicle,
                                      reference_location, other_location);
 
-      // Whichever vehicle's path is farthest away from the other vehicle gets
-      // priority to move.
-      if (cache.inter_geodesic_distance < 0.1 &&
-          ((cache.inter_bbox_distance > 0.1 &&
-            cache.reference_vehicle_to_other_geodesic > cache.other_vehicle_to_reference_geodesic
-            ) || (
-            cache.inter_bbox_distance < 0.1 &&
-            (cg::Math::Dot(reference_heading, reference_to_other) >
-            cg::Math::Dot(other_heading, other_to_reference))
-          ))
-      ) {
+      // Conditions for collision negotiation.
+      bool geodesic_path_bbox_touching = cache.inter_geodesic_distance < 0.1;
+      bool vehicle_bbox_touching = cache.inter_bbox_distance < 0.1;
+      bool ego_path_clear = cache.other_vehicle_to_reference_geodesic > 0.1;
+      bool other_path_clear = cache.reference_vehicle_to_other_geodesic > 0.1;
+      bool ego_path_priority = cache.reference_vehicle_to_other_geodesic < cache.other_vehicle_to_reference_geodesic;
+      bool ego_angular_priority = cg::Math::Dot(reference_heading, reference_to_other)
+                                  < cg::Math::Dot(other_heading, other_to_reference);
+
+      // Whichever vehicle's path is farthest away from the other vehicle gets priority to move.
+      if (geodesic_path_bbox_touching
+          && !(ego_path_clear && !other_path_clear)
+          && (!ego_path_clear
+              || (!vehicle_bbox_touching && !ego_path_priority)
+              || (vehicle_bbox_touching && !ego_angular_priority)))
+      {
 
         hazard = true;
+        // TODO: Remove thresholding after fixing parameters class to return clamped values.
+        const float specific_distance_margin = std::max(parameters.GetDistanceToLeadingVehicle(reference_vehicle),
+                                                        BOUNDARY_EXTENSION_MINIMUM);
+        available_distance_margin = static_cast<float>(std::max(cache.reference_vehicle_to_other_geodesic
+                                                                 - specific_distance_margin, 0.0));
```

Comment:
Instead of std::max use Macro 
#define MAX(__a, __b) ((__a) > (__b) ? (__a) : (__b))

Also compute values for __a and __b before sending them to macros.